// streams/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.asyncStream = exports.streamAsyncIterable = exports.asyncStreamIterable = void 0;
class AsyncStreamOfIterator {
    constructor(iterator) {
        this.iterator = iterator;
    }
    [Symbol.asyncIterator]() {
        return this.iterator;
    }
    filter(predicate) {
        async function* filtered(it) {
            for await (const v of it) {
                if (predicate(v)) {
                    yield v;
                }
            }
        }
        return new AsyncStreamOfIterator(filtered(this));
    }
    map(transform) {
        async function* mapped(it) {
            for await (const v of it) {
                yield transform(v);
            }
        }
        return new AsyncStreamOfIterator(mapped(this));
    }
    mapAwait(transform) {
        async function* mapAwaited(it) {
            for await (const v of it) {
                yield await transform(v);
            }
        }
        return new AsyncStreamOfIterator(mapAwaited(this));
    }
    flatMap(transform) {
        async function* flatMapped(it) {
            for await (const nested of it) {
                yield* transform(nested);
            }
        }
        return new AsyncStreamOfIterator(flatMapped(this));
    }
    flatMapAwait(transform) {
        async function* flatMapAwaited(it) {
            for await (const nested of it) {
                yield* await transform(nested);
            }
        }
        return new AsyncStreamOfIterator(flatMapAwaited(this));
    }
    batch(batchSize) {
        if (batchSize < 1) {
            throw new Error("batchSize should be positive");
        }
        async function* batched(it) {
            let acc = [];
            for await (const v of it) {
                acc.push(v);
                if (acc.length === batchSize) {
                    yield acc;
                    acc = [];
                }
            }
            if (acc.length > 0) {
                yield acc;
            }
        }
        return new AsyncStreamOfIterator(batched(this));
    }
    limit(maxSize) {
        async function* limited(it) {
            let count = 0;
            if (count >= maxSize) {
                return;
            }
            for await (const v of it) {
                yield v;
                count += 1;
                if (count >= maxSize) {
                    return;
                }
            }
        }
        return new AsyncStreamOfIterator(limited(this));
    }
    skip(n) {
        async function* skipped(it) {
            let count = 0;
            for await (const v of it) {
                if (count >= n) {
                    yield v;
                }
                count += 1;
            }
        }
        return new AsyncStreamOfIterator(skipped(this));
    }
    dropWhile(predicate) {
        async function* droppedWhile(it) {
            let dropping = true;
            for await (const v of it) {
                dropping = dropping && predicate(v);
                if (!dropping) {
                    yield v;
                }
            }
        }
        return new AsyncStreamOfIterator(droppedWhile(this));
    }
    takeWhile(predicate) {
        async function* takenWhile(it) {
            for await (const v of it) {
                if (!predicate(v)) {
                    return;
                }
                yield v;
            }
        }
        return new AsyncStreamOfIterator(takenWhile(this));
    }
    peek(observer) {
        async function* peeked(it) {
            for await (const v of it) {
                observer(v);
                yield v;
            }
        }
        return new AsyncStreamOfIterator(peeked(this));
    }
    async forEach(block) {
        for await (const v of this) {
            await block(v);
        }
    }
    async collect(container, accumulator, finisher) {
        for await (const v of this) {
            accumulator(container, v);
        }
        return finisher(container);
    }
    async reduceLeft(initial, reducer) {
        let result = initial;
        for await (const v of this) {
            result = reducer(result, v);
        }
        return result;
    }
    async all(predicate) {
        for await (const v of this) {
            if (!(await predicate(v))) {
                return false;
            }
        }
        return true;
    }
    async any(predicate) {
        for await (const v of this) {
            if (await predicate(v)) {
                return true;
            }
        }
        return false;
    }
    async none(predicate) {
        for await (const v of this) {
            if (await predicate(v)) {
                return false;
            }
        }
        return true;
    }
    async count() {
        let count = 0;
        for await (const v of this) {
            count += 1;
        }
        return count;
    }
    async first(predicate) {
        for await (const v of this) {
            if (await predicate(v)) {
                return v;
            }
        }
        return null;
    }
    async last(predicate) {
        let result = null;
        for await (const v of this) {
            if (await predicate(v)) {
                result = v;
            }
        }
        return result;
    }
    async max(comparator) {
        let result = null;
        for await (const v of this) {
            if (result === null) {
                result = v;
            }
            else {
                result = comparator(result, v) > 0 ? result : v;
            }
        }
        return result;
    }
    async min(comparator) {
        let result = null;
        for await (const v of this) {
            if (result === null) {
                result = v;
            }
            else {
                result = comparator(result, v) < 0 ? result : v;
            }
        }
        return result;
    }
    async reduce(adder) {
        let result = null;
        for await (const v of this) {
            if (result === null) {
                result = v;
            }
            else {
                result = adder(result, v);
            }
        }
        return result;
    }
    async toArray() {
        const result = [];
        for await (const v of this) {
            result.push(v);
        }
        return result;
    }
}
// Stream Sources
function asyncStreamIterable(itrbl) {
    async function* iterableSource() {
        for (const e of itrbl) {
            yield e;
        }
    }
    return new AsyncStreamOfIterator(iterableSource());
}
exports.asyncStreamIterable = asyncStreamIterable;
function streamAsyncIterable(itrbl) {
    return new AsyncStreamOfIterator(itrbl[Symbol.asyncIterator]());
}
exports.streamAsyncIterable = streamAsyncIterable;
function asyncStream(it) {
    if (typeof it[Symbol.iterator] === "function") {
        return asyncStreamIterable(it);
    }
    if (typeof it[Symbol.asyncIterator] === "function") {
        return streamAsyncIterable(it);
    }
    return new AsyncStreamOfIterator(it);
}
exports.asyncStream = asyncStream;
