// streams/readableAsyncIterator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.readableAsyncIterator = void 0;
function makePipe() {
    const puts = [];
    const takes = [];
    let closed = false;
    function put(n) {
        if (closed) {
            return;
        }
        const nextInLine = takes.shift();
        if (nextInLine) {
            nextInLine({ done: false, value: n });
        }
        else {
            puts.push(n);
        }
    }
    function close() {
        closed = true;
    }
    function next() {
        const next = puts.shift();
        if (next) {
            return Promise.resolve({ done: false, value: next });
        }
        if (closed) {
            return Promise.resolve({ done: true, value: undefined });
        }
        return new Promise(resolve => takes.push(resolve));
    }
    return { put, next, close };
}
function readableAsyncIterator(readable, by) {
    const { next, ...pipe } = makePipe();
    let remainder = by.initial();
    const dataListener = (chunk) => {
        let [items, nextRemainder] = by.split(chunk, remainder);
        items.forEach(pipe.put);
        remainder = nextRemainder;
    };
    const endListener = () => {
        const lastItem = by.last(remainder);
        if (lastItem !== null) {
            pipe.put(lastItem);
        }
        cleanUp();
    };
    readable.on("data", dataListener);
    readable.on("end", endListener);
    function cleanUp() {
        pipe.close();
        readable.removeListener("data", dataListener);
        readable.removeListener("end", endListener);
    }
    function iteratorReturn() {
        cleanUp();
        return Promise.resolve({ done: true, value: undefined });
    }
    return { next, return: iteratorReturn };
}
exports.readableAsyncIterator = readableAsyncIterator;
